#166791099 admin can update and delete book

Nennyfills opened the pull request:
### What does this PR do?
Add  update and delete book functionality

#### Description of Task to be completed?
- Create new endpoints to editBooks for admin
- Create new endpoints to deleteBooks  for admin
- Add swagger documentation
- Write unit test to test all routers

#### How should this be manually tested?
- Clone the project 
- Install all dependencies
- Run test
OR
- Run npm run db:rollback
- Start server and test manually with npm run start:dev
- Open postman use: 
     {put} for update
      1. `http://localhost:{PORT}/api/v1/admin/book:id`
      2. `http://localhost:{PORT}/api/v1/admin/book:id`
   {delete} for delete
      1. `http://localhost:{PORT}/api/v1/admin/book:id`
      2. `http://localhost:{PORT}/api/v1/admin/book:id`
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166791099](https://www.pivotaltracker.com/story/show/166791099)

#### Screenshots (if appropriate)?
![image](https://user-images.githubusercontent.com/45515619/59897131-29d20600-93e3-11e9-8635-246d9c664c0e.png)
![image](https://user-images.githubusercontent.com/45515619/59969186-07093400-9540-11e9-8d99-f50cbf7f3f1a.png)
![image](https://user-images.githubusercontent.com/45515619/59811311-2c5c2f00-9301-11e9-8946-b79444f4aaf6.png)
![image](https://user-images.githubusercontent.com/45515619/59811326-4433b300-9301-11e9-86b3-f47767bfb7c8.png)
![image](https://user-images.githubusercontent.com/45515619/59811338-56155600-9301-11e9-8eeb-4ce988a56650.png)

#### Questions:
N/A

allebd commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,55 @@ class BookController {
+      const getAuthor = await BookAuthor.findOne({ where: { bookId: +id }, attributes: ['authorId'] });
+      if (!books) return responseMessage(response, 400, { message: 'book does not exist' });
+      Book.update({
+        title, description, isbn, price, yearPublished
```
I observed that there is no condition to check if the same ISBN already exist because ISBN is meant to be unique

Nennyfills replied:
done

allebd commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,55 @@ class BookController {
+        title, description, isbn, price, yearPublished
+      }, { where: { id: +id } });
+      Author.update({ fullname: authorName }, { where: { id: getAuthor.authorId } });
+      return responseMessage(response, 200, { status: 'success', message: 'book successfully updated' });
```
I observed the **data** of the updated book is not in the response

Nennyfills replied:
done

allebd commented on server/middlewares/BookValidator.js:
```diff
@@ -164,4 +214,34 @@ export default class BookValidator {
+ * @returns {array} an array of check API middleware
+ * @memberof Validation
+ */
+  static EditValidation() {
```
The name EditValidation seems **ambiguous** can be renamed to either EditBookValidation or BookEditValidation

Nennyfills replied:
is already under the book validation class, there is no need to include book to it, the name is already long why make it longer

allebd commented on server/middlewares/BookValidator.js:
```diff
@@ -164,4 +214,34 @@ export default class BookValidator {
+ * @returns {array} an array of check API middleware
+ * @memberof Validation
+ */
+  static DeleteValidation() {
```
The name DeleteValidation seems  **ambiguous** too

Nennyfills replied:
is already under the book validation class, there is no need to include book to it, the name is already long why make it longer

Lundii reviewed (changes requested):
Good work so far. Here are things I would like you to look into

* I don't think there is a validator if the authorId field is omitted. This is what I got when I ommitted the authorId field
![Screenshot (100)](https://user-images.githubusercontent.com/37886814/59932546-5287e900-943f-11e9-8a59-772025007a6d.png)
The book book table is still updated even with the error above and when you try updating the book again, it tells you isbn already exits.

* There is no check if an author doesn't exist in the database before updating the BookAuthor table
* Adding `updatedAt: new Date()` when updating a book or bookauthor is not necessary as this gets done automatically.

Lundii commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,62 @@ class BookController {
+* @returns {JSON} object
+* @memberof BookController
+*/
+  static async editBook(request, response) {
```
`updateBook` sounds better

Nennyfills replied:
EditBook is okay too, is all about using a variable that will make someone understand what you're doing.

Lundii replied:
But you will agree that updateBook sounds a lot better

Nennyfills replied:
EditBook is also okay too..

henryade commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,62 @@ class BookController {
+    try {
+      const books = await Book.findAll();
+      const availableBook = books.find(book => book.id === +id);
+      if (!availableBook) return responseMessage(response, 400, { message: 'book does not exist' });
```
this should be 404 book not found

henryade commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,62 @@ class BookController {
+        if (availableIsbn === isbn) return responseMessage(response, 409, { message: 'isbn already exist' });
+      }
+      const results = await Book.update({
+        title, description, isbn, price, yearPublished, updatedAt: new Date()
```
remove all updateAt

Nennyfills commented:
> Good work so far. Here are things I would like you to look into
> 
> * I don't think there is a validator if the authorId field is omitted. This is what I got when I ommitted the authorId field
>   ![Screenshot (100)](https://user-images.githubusercontent.com/37886814/59932546-5287e900-943f-11e9-8a59-772025007a6d.png)
>   The book book table is still updated even with the error above and when you try updating the book again, it tells you isbn already exits.
> * There is no check if an author doesn't exist in the database before updating the BookAuthor table
> * Adding `updatedAt: new Date()` when updating a book or bookauthor is not necessary as this gets done automatically.

done

henryade commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,62 @@ class BookController {
+      const author = await BookAuthor.update({ authorId, updatedAt: new Date() }, {
+        where: { bookId: +id }, returning: true, raw: true
+      });
+      const newAuthorId = author[1][0].authorId;
```
this line is redundant

henryade commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,62 @@ class BookController {
+    const { id } = request.params;
+    try {
+      const books = await Book.findOne({ where: { id: +id } });
+      if (!books) return responseMessage(response, 400, { message: 'book does not exist' });
```
404 also

Nennyfills replied:
I was told to leave it as 400

JuwonAbiola reviewed (commented):
Good job so far this is what noticed when i passed an empty request body 
![image](https://user-images.githubusercontent.com/42170732/59947560-71997180-9465-11e9-9dd1-fb6ec529a502.png)

please make use of the emptyBody function in your validator you can find it in the helper folder

Lundii commented:
> Good job so far this is what noticed when i passed an empty request body
> ![image](https://user-images.githubusercontent.com/42170732/59947560-71997180-9465-11e9-9dd1-fb6ec529a502.png)
> 
> please make use of the emptyBody function in your validator you can find it in the helper folder

* In addition to this, this is what I got when I tried to update just the authorId
![Screenshot (101)](https://user-images.githubusercontent.com/37886814/59949610-8b3db780-946b-11e9-878c-7f24bb2fa9a4.png)
* You should also update your PR screenshots as some of them is having authorName and others authorId

Lundii commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,65 @@ class BookController {
+      }
+      const results = await Book.update({
+        title, description, isbn, price, yearPublished
+      }, { where: { id: +id }, returning: true, plain: true });
```
The plus (+) sign in front of id is not really necessary though. It would would without it,

Nennyfills replied:
is necessary. without it the id is a string

Ukhu commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,65 @@ class BookController {
+      });
+      const author = affectedRows;
+      return responseMessage(response, 200, {
+        status: 'success', message: 'book successfully updated', results: { ...results, author }
```
In the response message on your screenshot, the results object in this line isn't showing

Nennyfills replied:
please check well

Ukhu replied:
As at when I made this comment 2 days ago, it wasn't there. Bella also noticed the same thing. You would simply have said that you have fixed it.

Ukhu commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,65 @@ class BookController {
+      title, description, isbn, price, yearPublished, authorId
+    } = request.body;
+    try {
+      const books = await Book.findAll();
```
If the database has a large number of books, this query here will be very intensive for the server. Why not query the database using the 'findOne' method and a filter to get 'availableBook' and 'availableIsbn'? That would be better

Ukhu commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,65 @@ class BookController {
+        if (availableIsbn === isbn) return responseMessage(response, 409, { message: 'isbn already exist' });
+      }
+      const results = await Book.update({
+        title, description, isbn, price, yearPublished
```
Assuming the admin only inputs the title in the request body, and the title is the exact same thing as what was previously there, from your logic in this controller, the response would be 'book successfully updated'. Should that be the correct response, even though there was no update?

Nennyfills replied:
once you have requested for it, it is marked as updated

allebd reviewed (commented):
Kindly update your screenshot for a successful book updated in your PR

Nennyfills commented:
> > Good job so far this is what noticed when i passed an empty request body
> > ![image](https://user-images.githubusercontent.com/42170732/59947560-71997180-9465-11e9-9dd1-fb6ec529a502.png)
> > please make use of the emptyBody function in your validator you can find it in the helper folder
> 
> In addition to this, this is what I got when I tried to update just the authorId
> ![Screenshot (101)](https://user-images.githubusercontent.com/37886814/59949610-8b3db780-946b-11e9-878c-7f24bb2fa9a4.png)

you only updated  authorId, and it gave you the result of the id you updated, the 0:0 means you updated nothing in the book database

Ukhu reviewed (commented):
In the server response for when a book is successfully updated, can you change that 'results' to 'data' for consistency with the other endpoints. Secondly, instead of having a nested object inside the results object, kindly format it so that the results(or data) object will contain both the details of the updated book and the new author (if the author was changed) as it properties.

Nennyfills commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,66 @@ class BookController {
+      if (authorId) {
+        const existingAuthorId = await Author.findOne({ where: { id: authorId } });
+        if (!existingAuthorId) return responseMessage(response, 400, { message: 'author does not exist' });
+        const result = await BookAuthor.update({ authorId },
```
eslint didn't flag that.

> Expected a newline after '(' function-paren-newline

eslint didn't flag that.

Nennyfills commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,66 @@ class BookController {
+        const existingAuthorId = await Author.findOne({ where: { id: authorId } });
+        if (!existingAuthorId) return responseMessage(response, 400, { message: 'author does not exist' });
+        const result = await BookAuthor.update({ authorId },
+          { where: { bookId: +id }, returning: true, raw: true });
```
> Expected a newline before ')' function-paren-newline

is not needed, I recommend you check this locally, there you will see that eslint didn't flag that.

Ukhu replied:
You are asking Hound CI to check your code locally?

Nennyfills commented:
> In the server response for when a book is successfully updated, can you change that 'results' to 'data' for consistency with the other endpoints. Secondly, instead of having a nested object inside the results object, kindly format it so that the results(or data) object will contain both the details of the updated book and the new author (if the author was changed) as it properties.

am returning it from the affected row in the update function, which comes as an array. i have already destructured it.  check  the code you will understand what am saying

Nennyfills commented:
> Kindly update your screenshot for a successful book updated in your PR
please check well is already there.

Ukhu commented:
> > In the server response for when a book is successfully updated, can you change that 'results' to 'data' for consistency with the other endpoints. Secondly, instead of having a nested object inside the results object, kindly format it so that the results(or data) object will contain both the details of the updated book and the new author (if the author was changed) as it properties.
> 
> am returning it from the affected row in the update function, which comes as an array. i have already destructured it. check the code you will understand what am saying

You just worked on that after I made my comment because your screenshot isn't showing 'results' again, it's now showing 'data' as I suggested. Once again, you would have simply said that you have worked on it.

Ukhu commented on server/controllers/BookController.js:
```diff
@@ -75,6 +75,70 @@ class BookController {
+    const {
+      title, description, isbn, price, yearPublished, authorId
+    } = request.body;
+    // const bodyData = title || description || isbn || price || yearPublished;
```
remove this commented code from here since you are not making use of it.

Nennyfills replied:
this is not commented .... have no idea why you have it commented, will recommit and push again

Ukhu reviewed (changes requested):
When I tried updating only the title of a book, the server returned an empty data object.

![Screenshot (179)](https://user-images.githubusercontent.com/35268825/59974902-50419e00-95a9-11e9-94ed-00be90b0b8aa.png)

Nennyfills commented:
> > > In the server response for when a book is successfully updated, can you change that 'results' to 'data' for consistency with the other endpoints. Secondly, instead of having a nested object inside the results object, kindly format it so that the results(or data) object will contain both the details of the updated book and the new author (if the author was changed) as it properties.
> > 
> > 
> > am returning it from the affected row in the update function, which comes as an array. i have already destructured it. check the code you will understand what am saying
> 
> You just worked on that after I made my comment because your screenshot isn't showing 'results' again, it's now showing 'data' as I suggested. Once again, you would have simply said that you have worked on it.

this happened because of the above-commented code

Nennyfills commented:
> > > In the server response for when a book is successfully updated, can you change that 'results' to 'data' for consistency with the other endpoints. Secondly, instead of having a nested object inside the results object, kindly format it so that the results(or data) object will contain both the details of the updated book and the new author (if the author was changed) as it properties.
> > 
> > 
> > am returning it from the affected row in the update function, which comes as an array. I have already destructured it. check the code you will understand what am saying
> 
> You just worked on that after I made my comment because your screenshot isn't showing 'results' again, it's now showing 'data' as I suggested. Once again, you would have simply said that you have worked on it.

you're getting me wrong, I only worked on changing the result to data, not on this restructuring.

Nennyfills commented:
> When I tried updating only the title of a book, the server returned an empty data object.
> 
> ![Screenshot (179)](https://user-images.githubusercontent.com/35268825/59974902-50419e00-95a9-11e9-94ed-00be90b0b8aa.png)

done

Nennyfills commented:
> In the server response for when a book is successfully updated, can you change that 'results' to 'data' for consistency with the other endpoints. Secondly, instead of having a nested object inside the results object, kindly format it so that the results(or data) object will contain both the details of the updated book and the new author (if the author was changed) as it properties.

done

Nennyfills committed:
166791099-story(admin): update and delete book
 - create new endpoints to editBooks for admin
 - create new endpoints to deleteBooks  for admin
 - add swagger documentation
 - write unit test to test all routers
[Started  #166791099]
